mlir/lib/Quantum/Transforms/wrapper_for_wrapper.cpp

```diff
@@ -0,0 +1,220 @@
+    return result;
+}
+
+static bool functionHasInputs(LLVM::LLVMFuncOp op)
+{
+    bool result = false;
+    op.walk([&](LLVM::CallOp callOp) { result = !callOp.getOperandTypes().empty(); });
+    return result;
+}
+
+static LLVM::LLVMFunctionType
+convertFunctionTypeCatalystWrapper(PatternRewriter &rewriter, LLVM::LLVMFunctionType functionType,
+                                   bool hasReturns, bool hasInputs)
+{
+    SmallVector<Type, 2> transformedInputs;
+
+    Type ptr = LLVM::LLVMPointerType::get(rewriter.getContext());
+    auto inputs = functionType.getParams();
+    Type resultType = hasReturns ? inputs.front() : ptr;
+    transformedInputs.push_back(resultType);
+
+    if (hasReturns) {
+        inputs = inputs.drop_front();
+    }
+
+    LLVMTypeConverter typeConverter(rewriter.getContext());
+    Type inputType = hasInputs ? typeConverter.packFunctionResults(inputs) : ptr;
+    bool noChange = inputs.size() == 1;
+    if (noChange) {
+        // Still wrap the pointer into a struct
+        // for uniformity in Python and in the unwrapping.
+        inputType = LLVM::LLVMStructType::getLiteral(rewriter.getContext(), inputType);
+    }
+    if (auto structType = inputType.dyn_cast<LLVM::LLVMStructType>()) {
+        inputType = LLVM::LLVMPointerType::get(structType);
+    }
+    transformedInputs.push_back(inputType);
+
+    Type voidType = LLVM::LLVMVoidType::get(rewriter.getContext());
+    return LLVM::LLVMFunctionType::get(voidType, transformedInputs);
+}
+
+static void wrapResultsAndArgsInTwoStructs(LLVM::LLVMFuncOp op, PatternRewriter &rewriter,
+                                           std::string nameWithoutPrefix)
+{
+    bool hasReturns = functionHasReturns(op);
+    bool hasInputs = functionHasInputs(op);
+    LLVM::LLVMFunctionType functionType = op.getFunctionType();
+    LLVM::LLVMFunctionType wrapperFuncType =
+        convertFunctionTypeCatalystWrapper(rewriter, functionType, hasReturns, hasInputs);
+
+    Location loc = op.getLoc();
+    auto wrapperFuncOp = rewriter.create<LLVM::LLVMFuncOp>(
+        loc, llvm::formatv("_catalyst_pyface_{0}", nameWithoutPrefix).str(), wrapperFuncType,
+        LLVM::Linkage::External, /*dsoLocal*/ false,
+        /*cconv*/ LLVM::CConv::C);
+
+    OpBuilder::InsertionGuard guard(rewriter);
+    rewriter.setInsertionPointToStart(wrapperFuncOp.addEntryBlock());
+
+    auto type = op.getFunctionType();
+    auto params = type.getParams();
+    SmallVector<Value, 8> args;
+
+    if (hasReturns) {
+        args.push_back(wrapperFuncOp.getArgument(0));
+        params = params.drop_front();
+    }
+
+    if (hasInputs) {
+        Value arg = wrapperFuncOp.getArgument(1);
+        Value structOfMemrefs = rewriter.create<LLVM::LoadOp>(loc, arg);
+
+        for (auto &en : llvm::enumerate(params)) {
+            Value pointer = rewriter.create<LLVM::ExtractValueOp>(loc, structOfMemrefs, en.index());
+            args.push_back(pointer);
```

Comment:
Are these pointers or just memrefs?

Reply:
Memref might be too high level to describe "memref"s at this level of abstraction. During LLVM-IR, there are no memrefs, just struct types, pointer types and primitive types. It just so happens that memrefs get lowered to a struct type. The value held in `pointer` is indeed a pointer to a struct which would correspond to the lowering of a memref.